inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/pooling/pooling_kernel_gpu_b_fs_yx_fsv16_imad.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "pooling_kernel_gpu_b_fs_yx_fsv16_imad.h"
+#include "kernel_selector_utils.h"
+
+#define FEATURE_SLICE_SIZE 16
+
+namespace kernel_selector {
+ParamsKey PoolingKernelGPU_b_fs_yx_fsv16_imad::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::INT8);
+    k.EnableInputDataType(Datatype::UINT8);
+    k.EnableOutputDataType(Datatype::INT8);
+    k.EnableOutputDataType(Datatype::UINT8);
+    k.EnableOutputDataType(Datatype::F32);
+    k.EnableInputLayout(DataLayout::b_fs_yx_fsv16);
+    k.EnableOutputLayout(DataLayout::b_fs_yx_fsv16);
+    k.EnableTensorOffset();
+    k.EnableTensorPitches();
+    k.EnableBatching();
+    k.EnablePoolType(PoolType::MAX);
+    k.EnablePoolType(PoolType::AVG);
+    k.EnablePoolRemainder(PoolRemainder::FLOOR);
+    k.EnablePoolRemainder(PoolRemainder::CEIL);
+    k.EnablePoolKernelDividerMode(KernelDividerMode::FIXED);
+    k.EnablePoolKernelDividerMode(KernelDividerMode::DYNAMIC);
+    k.EnablePoolKernelDividerMode(KernelDividerMode::DYNAMIC_WITH_PADDING);
+    k.EnableDifferentTypes();
+    return k;
+}
+
+PoolingKernelBase::DispatchData PoolingKernelGPU_b_fs_yx_fsv16_imad::SetDefault(const pooling_params& params) const {
+    DispatchData runInfo = PoolingKernelBase::SetDefault(params);
+
+    const auto& out = params.output;
+    auto x = out.X().v;
+    auto y = out.Y().v;
+    auto f = out.Feature().v;
+    auto b = out.Batch().v;
+
+    runInfo.gws0 = x;
+    runInfo.gws1 = y;
+    // we got b_fs_yx_fsv16 format, we process 16 features per workitem
+    runInfo.gws2 = CeilDiv(f, FEATURE_SLICE_SIZE) * b;
```

Comment:
Have you tried to enforce subgroup size = 16 and use block reads?

Reply:
I'm not sure if I understand correctly but if you mean setting lws size to 16 then it doesn't work if gws is less than 16 and I'm not sure also how can I use block read in kernel in this case.

Reply:
I'm not sure if understand correctly but if you mean setting lws to 16 then there is a problem and it doesn't work if gws is less than 16. Also I'm not sure how can I use block read in kernel in these case.

Reply:
@mkaczyns 

```cpp
runInfo.gws2 = Align(f, FEATURE_SLICE_SIZE) * b;
runInfo.lws2 = FEATURE_SLICE_SIZE;
```
and then set attribute for the kernel:
```
__attribute__((intel_reqd_sub_group_size(FEATURE_SLICE_SIZE)))
```

Reply:
@mkaczyns @vladimir-paramuzov 
Changing this kernel to use sub-groups will also require major changes in algorithm, simple replacement will result in worse performance (any simd -> simd 16, vload 4 x 4bytes -> block read 1 x 1 byte).
This kernel additionally has exactly the same algorithm and parallelization scheme as fsv4 pooling.
My proposition is to generalize existing fsv4 kernel and leave possible sub-group optimization as a separate task if current approach will not be efficient enough.

Reply:
@kdobros  @mkaczyns  OK, I checked the asm for this kernel, the compiler generated single send + a lot of extra stuff for data shuffle, which is not bad from the performance perspective.
So we can merge this PR once all other comments are addressed, but I still think that the experiment with blocked load is needed to have clear picture of the optimization level of this kernel and estimate a possibility of further improvements